File: fn/src/main/java/org/sqlbuilder/fn/joins/FE_FEExcluded.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

import edu.berkeley.icsi.framenet.InternalFrameRelationFEType;

public class FE_FEExcluded extends Pair<Integer, Integer> implements Insertable
{
	public static final Set<FE_FEExcluded> SET = new HashSet<>();

	// C O N S T R U C T O R

	public static FE_FEExcluded make(final int feid, final InternalFrameRelationFEType fe2)
	{
		var ff = new FE_FEExcluded(feid, fe2.getID());
		SET.add(ff);
		return ff;
	}

	private FE_FEExcluded(final int feid, final int feid2)
	{
		super(feid, feid2);
	}

	// O R D E R

	public static final Comparator<FE_FEExcluded> COMPARATOR = Comparator.comparing(FE_FEExcluded::getFirst).thenComparing(FE_FEExcluded::getSecond);

	// I N S E R T

	@Override
	public String dataRow()
	{
		return String.format("%d,%d", first, second);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[FE-FEexcl feid=%s feid2=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/FE_FEExcluded.kt
package org.sqlbuilder.fn.joins

import edu.berkeley.icsi.framenet.InternalFrameRelationFEType
import org.sqlbuilder.common.Insertable
import java.util.Objects

data class FE_FEExcluded(
    val feid: Int,
    val feid2: Int,
) : Insertable {

    // I N S E R T

    override fun dataRow(): String {
        return "$feid,$feid2"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as FE_FEExcluded
        return feid == that.feid && feid2 == that.feid2
    }

    override fun hashCode(): Int {
        return Objects.hash(feid, feid2)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[FE-exclFE feid=$feid feid2=$feid2]"
    }

    companion object {

        @JvmField
        val COMPARATOR: Comparator<FE_FEExcluded> = Comparator
            .comparing<FE_FEExcluded, Int> { it.feid }
            .thenComparing<Int?> { it.feid2 }

        @JvmField
        val SET = HashSet<FE_FEExcluded>()

        fun make(feid: Int, fe2: InternalFrameRelationFEType): FE_FEExcluded {
            val ff = FE_FEExcluded(feid, fe2.getID())
            SET.add(ff)
            return ff
        }
    }
}
